src/libraries/Microsoft.PowerFx.Core/Public/Config/ReadOnlySymbolTable.cs

```
@@ -0,0 +1,300 @@
+            };
+
+            foreach (var func in coreFunctions)
+            {
+                s.AddFunction(func); // will also add enum. 
+            }
+
+            return s;
+        }
+
+        private protected readonly Dictionary<string, NameLookupInfo> _variables = new Dictionary<string, NameLookupInfo>();
+
+        private protected readonly List<TexlFunction> _functions = new List<TexlFunction>();
+
+        // Which enums are available. 
+        // These do not compose - only bottom one wins. 
+        private protected EnumStoreBuilder _enumStoreBuilder;
+        private EnumSymbol[] _enumSymbolCache;
+
+        private EnumSymbol[] GetEnumSymbolSnapshot
+        {
+            get
+            {
+                // The caller may add to the builder after we've assigned. 
+                // So delay snapshot until we actually need to read it. 
+                if (_enumStoreBuilder == null)
+                {
+                    _enumSymbolCache = new EnumSymbol[] { };
+                }
+
+                if (_enumSymbolCache == null)
+                {
+                    _enumSymbolCache = _enumStoreBuilder.Build().EnumSymbols.ToArray();
+                }
+
+                return _enumSymbolCache;
+            }
+        }
+
+        internal IEnumerable<TexlFunction> Functions => ((INameResolver)this).Functions;
+
+        IEnumerable<TexlFunction> INameResolver.Functions => _functions;
+
+        IExternalDocument INameResolver.Document => default;
+
+        IExternalEntityScope INameResolver.EntityScope => throw new NotImplementedException();
+
+        IExternalEntity INameResolver.CurrentEntity => default;
+
+        DName INameResolver.CurrentProperty => default;
+
+        DPath INameResolver.CurrentEntityPath => default;
+
+        bool INameResolver.SuggestUnqualifiedEnums => false;
+
+        internal readonly Dictionary<DName, IExternalEntity> _environmentSymbols = new Dictionary<DName, IExternalEntity>();
```

small readability thought:
Can we move all the default/notimpl left overs to the end/put it in a  region to hide or something like that?

---

yes!